share/get_proof.go

```diff
@@ -0,0 +1,52 @@
+package share
+
+import (
+	"context"
+
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+
+	"github.com/celestiaorg/celestia-node/share/ipld"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// SharesWithProofs contains data with corresponding Merkle Proof
+type SharesWithProofs struct {
+	// Shares are the full data including namespace
+	Shares []Share
+	// Proof is a Merkle Proof for the corresponding Shares
+	Proof nmt.Proof
+}
+
+// GetSharesWithProofsByNamespace walks the tree of a given root and returns its shares within the
+// given namespace.ID. If a share could not be retrieved, err is not nil, and the returned array
+// contains nil shares in place of the shares it was unable to retrieve.
+func GetSharesWithProofsByNamespace(
```

Comment:
Why not follow the same approach by passing nmt.Proof optionally?

Reply:
Yeah, it is better to update this func too. But I'd opt for using `*ipld.Proof` instead, since it could be encoded for wire transfer and `nmt.Proof` doesn't have such option.